server/src/main.rs

```diff
@@ -0,0 +1,156 @@
+                if public_key.verify(&payload.as_bytes(), &signature_bytes).is_ok() {
+                    return true
+                }
+            }
+        }
+    }
+    return false
+}
+
+fn respond_success (is_verified: bool) -> hyper::Response<Body> {
+    // construct response based on verification status
+    match is_verified {
+        true => {
+            Response::builder()
+                .status(StatusCode::OK)
+                .body(Body::empty())
+                .unwrap()
+        },
+        _ => {
+            Response::builder()
+                .status(StatusCode::UNAUTHORIZED)
+                .body(Body::empty())
+                .unwrap()
+        }
+    }
+}
+
+fn read_hp_pubkey() -> PublicKey {
+    let hpos_state_path = env::var("HPOS_STATE_PATH").expect("HPOS_STATE_PATH environmental variable is not present");
+
+    println!("Reading HP Admin Public Key from {}.", hpos_state_path);
+
+    // Read from path
+    let contents = fs::read(hpos_state_path)
+        .expect("Something went wrong reading HP Admin Public Key from file");
+
+    // Parse content
+    let hpos_state: State = serde_json::from_slice(&contents).unwrap();
+    hpos_state.get_admin_public_key().expect("HP Admin Public key seems to be corrupted")
+}
+
+fn main() {
+    // Listen on http socket port 2884 - "auth" in phonespell
+    let listen_address = ([127,0,0,1], 2884).into();
+
+    // Trigger lazy static to see if HP_PUBLIC_KEY assignment creates panic
+    let _ = &*HP_PUBLIC_KEY;
+
+    // Create a `Service` from servicing function
+    let new_svc = || {
+        service::service_fn(create_response)
+    };
+
+    let server = Server::bind(&listen_address)
+        .serve(new_svc)
+        .map_err(|e| {
+            eprintln!("server error: {}", e);
+        });
+
+    println!("Listening on http://{}", listen_address);
+
+    // Run forever
+    hyper::rt::run(server);
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use ed25519_dalek;
+    use std::convert::From;
+
+    #[test]
+    fn verify_request_smoke() {
+        // Get a legit request_hash signature, agent_id
+        let secret: [u8; 32] = [0_u8; 32];
+        let secret_key = ed25519_dalek::SecretKey::from_bytes( &secret ).unwrap();
+        let public_key = ed25519_dalek::PublicKey::from( &secret_key );
+        let secret_key_exp = ed25519_dalek::ExpandedSecretKey::from( &secret_key );
+
+        // Now lets sign some body payload
+        let body = json!({
+            "something": "interesting"
+        });
+        let body_json = serde_json::to_string( &body ).unwrap();
+
+        let signature = secret_key_exp.sign( body_json.as_bytes(), &public_key );
+        let mut signature_base64 = String::new();
+        base64::encode_config_buf(signature.to_bytes().as_ref(), base64::STANDARD, &mut signature_base64);
+
+        let mut headers = HeaderMap::new();
+        headers.insert( "x-hpos-admin-signature", signature_base64.parse().unwrap() );
```

Comment:
```suggestion
        headers.insert(X_HPOS_ADMIN_SIGNATURE, signature_base64.parse().unwrap() );
```

Reply:
Tried that, couldn't figure out how to make it work...

Reply:
I was able to get this project to build in nix, when I run it, I get the error:

```
result/bin/hp-admin-crypto-server 
thread 'main' panicked at 'HPOS_STATE_PATH environmental variable is not present: NotPresent', src/libcore/result.rs:1051:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
```

I was able to add this env var in the development context (for local machine development) to the shell.nix file like this

```
{ pkgs ? import ./pkgs.nix {} }:

with pkgs;

mkShell {
  inputsFrom = lib.attrValues (import ./. { inherit pkgs; });
  HPOS_STATE_PATH = "blah";
}
```

then running `nix-shell`

the variable is found 

```
 /nix/store/07cqdikifgsn0krv30ds3x2dc1rhvg9v-hp-admin-authorize/bin/hp-admin-crypto-server
Reading HP Admin Public Key from blah.
thread 'main' panicked at 'Something went wrong reading HP Admin Public Key from file: Os { code: 2, kind: NotFound, message: "No such file or directory" }', src/libcore/result.rs:1051:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
```

```

Reply:
I pushed this to PR #3